tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp

```diff
@@ -1906,9 +1932,130 @@ void SwiftLangSupport::getCursorInfo(
+        getASTManager()->getMemoryBuffer(RealInputFilePath, fileSystem,
+                                         InputFileError);
+    if (UnresolvedInputFile) {
+      auto SolverBasedReceiver = [&](const RequestResult<CursorInfoData> &Res) {
+        SolverBasedResultCount = ResultCount(Res);
+        SolverBasedResultDescription = ResultDescription(Res);
+      };
+
+      CompilerInvocation CompInvok;
+      Invok->applyTo(CompInvok);
+
+      performWithParamsToCompletionLikeOperation(
+          UnresolvedInputFile.get(), Offset,
+          /*InsertCodeCompletionToken=*/false, Args, fileSystem,
+          CancellationToken,
+          [&](CancellableResult<CompletionLikeOperationParams> ParmsResult) {
+            ParmsResult.mapAsync<CursorInfoResults>(
+                [&](auto &Params, auto DeliverTransformed) {
+                  getCompletionInstance()->cursorInfo(
+                      Params.Invocation, Args, fileSystem,
+                      Params.completionBuffer, Offset, Params.DiagC,
+                      Params.CancellationFlag, DeliverTransformed);
+                },
+                [&](auto Result) {
+                  deliverCursorInfoResults(SolverBasedReceiver, Result, *this,
+                                           CompInvok, Actionables, SymbolGraph);
+                });
+          });
+    }
+  }
+
+  auto VerifySolverBasedResult =
+      [ResultCount, ResultDescription, SolverBasedResultCount,
+       SolverBasedResultDescription](
+          const RequestResult<CursorInfoData> &ASTBasedResult) {
+        if (SolverBasedResultDescription.empty()) {
+          // We did not run the solver-based implementation. Nothing to check.
+          return;
+        }
+        auto ASTResultDescription = ResultDescription(ASTBasedResult);
+        auto ASTResultCount = ResultCount(ASTBasedResult);
+        if (ASTResultCount == 0 && SolverBasedResultCount > 0) {
+          // The AST-based implementation did not return any results but the
+          // solver-based did. That's an improvement. Success.
+          return;
+        }
+        if (SolverBasedResultDescription == ASTResultDescription) {
+          // The solver-based and AST-based implementation produced the same
+          // results. Success.
+          return;
+        }
+        // The solver-based implementation differed from the AST-based
+        // implementation. Report a failure.
+        llvm::errs() << SolverBasedResultDescription << "\n";
+        llvm::errs() << "===== (solver based vs legacy) =====\n";
+        llvm::errs() << ASTResultDescription << "\n";
+        assert(false);
+      };
+
+  // Thunk around `Receiver` that, if solver-based cursor info is enabled,
+  // verifies that the solver-based cursor info result matches the AST-based
+  // result.
+  auto ReceiverThunk = [Receiver, VerifySolverBasedResult](
+                           const RequestResult<CursorInfoData> &Res) {
+#ifndef NDEBUG
```

Comment:
Actually, I think that check is not necessary at all because `VerifySolverBasedResult` will just return if `SolverBasedResultDescription` is empty. I’ll just remove it.